CMakeLists.txt

```diff
@@ -1,52 +1,65 @@
 cmake_minimum_required(VERSION 3.14)
+
+# The C++ version of PyTorch is also known as libtorch.
 project(PyTorch)
-## libtorch requires C++-14
+
+# PyTorch requires C++-14.  It requires gcc-8 or earlier.
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
-if(APPLE)
-  set(torch_fetch_URL "https://download.pytorch.org/libtorch/cpu/libtorch-macos-1.5.1.zip")
-  set(torch_fetch_HASH "MD5=77ca8e871c1200f1e5864ae1d9251a48")
-elseif(WIN32)
-  set(torch_fetch_URL "https://download.pytorch.org/libtorch/cu101/libtorch-win-shared-with-deps-1.5.1%2Bcu101.zip")
-  set(torch_fetch_HASH "MD5=90eb7739fc3b2f4572a93ee6da1e2139")
-else()
-  set(torch_fetch_URL "https://download.pytorch.org/libtorch/cu101/libtorch-cxx11-abi-shared-with-deps-1.5.1%2Bcu101.zip")
-  set(torch_fetch_HASH "MD5=b18d6f2fdfe730c1fca3edd5ed2f3c8f")
-endif()
+# We will need to use the location ${Python_EXECUTABLE}
+find_package(Python REQUIRED)
 
+# Get PyTorch and its dependencies, git recursively
+set(torch_fetch_git_repository "https://github.com/pytorch/pytorch")
+set(torch_fetch_git_tag "v1.5.1")
 include(FetchContent)
 FetchContent_Declare(
   torch_fetch
-  URL           "${torch_fetch_URL}"
-  URL_HASH      "${torch_fetch_HASH}"
+  GIT_REPOSITORY "${torch_fetch_git_repository}"
+  GIT_TAG        "${torch_fetch_git_tag}"
 )
+message(STATUS "FetchContent_Populate ${torch_fetch_git_repository} ${torch_fetch_git_tag}")
 FetchContent_Populate(torch_fetch)
+message(STATUS "FetchContent_Populate ${torch_fetch_git_repository} ${torch_fetch_git_tag} - done")
+
+# Use torch's setup.py to build and locally install it.
+if(NOT DEFINED torch_fetch_install_directory)
+  # We can enable and disable functionality with the following options.
+  if(NOT DEFINED ENV{USE_CUDA})
+    # We will not use CUDA on GitHub
+    set(ENV{USE_CUDA} "0")
+    set(torch_fetch_additional_libraries "torch_cuda" "c10_cuda")
+  endif()
+  if(NOT DEFINED ENV{USE_CUDNN})
+    # We will not use CUDNN on GitHub
+    set(ENV{USE_CUDNN} "0")
+  endif()
+  set(torch_fetch_install_directory "${torch_fetch_SOURCE_DIR}/../torch_fetch-install")
+  set(torch_fetch_build_directory "${torch_fetch_SOURCE_DIR}/build")
+  message(STATUS "Executing: cd ${torch_fetch_SOURCE_DIR}; ${Python_EXECUTABLE} setup.py install --prefix=${torch_fetch_install_directory}")
+  add_custom_target(
```

Comment:
I think this is executed at build time, not at compile time, so surrounding messages are misleading.

Reply:
I talked with @thewtex about how to revamp the `CMakeLists.txt` file.  I will make those changes in a different pull request.